src/audio/chain_dma.c

```diff
@@ -333,6 +240,114 @@ static enum task_state chain_task_run(void *data)
 	return SOF_TASK_STATE_RESCHEDULE;
 }
 
+static int chain_task_start(struct comp_dev *dev)
+{
+	struct comp_driver_list *drivers = comp_drivers_get();
+	struct chain_dma_data *cd = comp_get_drvdata(dev);
+	k_spinlock_key_t key;
+	int ret;
+
+	comp_info(dev, "chain_task_start(), host_dma_id = 0x%08x", cd->host_connector_node_id.dw);
+
+	key = k_spin_lock(&drivers->lock);
+	switch (cd->chain_task.state) {
+	case SOF_TASK_STATE_QUEUED:
+		k_spin_unlock(&drivers->lock, key);
+		return 0;
+	case SOF_TASK_STATE_COMPLETED:
+		break;
+	case SOF_TASK_STATE_INIT:
+		break;
+	case SOF_TASK_STATE_FREE:
+		break;
+	default:
+		comp_err(dev, "chain_task_start(), bad state transition");
+		ret = -EINVAL;
+		goto error;
+	}
+
+	if (cd->stream_direction == SOF_IPC_STREAM_PLAYBACK) {
+		ret = chain_host_start(dev);
+		if (ret)
+			goto error;
+		ret = chain_link_start(dev);
+		if (ret) {
+			chain_host_stop(dev);
+			goto error;
+		}
+	} else {
+		ret = chain_link_start(dev);
+		if (ret)
+			goto error;
+		ret = chain_host_start(dev);
+		if (ret) {
+			chain_link_stop(dev);
+			goto error;
+		}
+	}
+
+	pm_policy_state_lock_get(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);
+
+	ret = schedule_task_init_ll(&cd->chain_task, SOF_UUID(chain_dma_uuid),
+				    SOF_SCHEDULE_LL_TIMER, SOF_TASK_PRI_HIGH,
+				    chain_task_run, cd, 0, 0);
+	if (ret < 0) {
+		comp_err(dev, "chain_task_start(), ll task initialization failed");
+		rfree(&cd->chain_task);
+		goto error_task;
+	}
+
+	ret = schedule_task(&cd->chain_task, 0, 0);
+	if (ret)
+		goto error_task;
+
+	cd->chain_task.state = SOF_TASK_STATE_INIT;
+	k_spin_unlock(&drivers->lock, key);
+
+	return 0;
+
+error_task:
+	pm_policy_state_lock_put(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);
+	chain_host_stop(dev);
+	chain_link_stop(dev);
+error:
+	k_spin_unlock(&drivers->lock, key);
+	return ret;
+}
+
+static int chain_task_pause(struct comp_dev *dev)
+{
+	struct comp_driver_list *drivers = comp_drivers_get();
+	struct chain_dma_data *cd = comp_get_drvdata(dev);
+	k_spinlock_key_t key;
+	int ret, ret2;
+
+	if (cd->chain_task.state == SOF_TASK_STATE_FREE)
+		return 0;
+
+	key = k_spin_lock(&drivers->lock);
+	cd->first_data_received = false;
+	if (cd->stream_direction == SOF_IPC_STREAM_PLAYBACK) {
+		ret = chain_host_stop(dev);
+		ret2 = chain_link_stop(dev);
+	} else {
+		ret = chain_link_stop(dev);
+		ret2 = chain_host_stop(dev);
+	}
+	if (!ret)
+		ret = ret2;
+	if (ret < 0)
+		goto error;
+
+	cd->chain_task.state = SOF_TASK_STATE_COMPLETED;
+	schedule_task_free(&cd->chain_task);
+
+error:
+	pm_policy_state_lock_put(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);
```

Comment:
so in error case you don't free the task, but you do put the PM policy state lock, what's the logic behind this? BTW, reviewing this would've been easier is you kept functions in place, then differences could be seen directly

Reply:
Moving `schedule_task_init` to `chain_task_start()` caused some undesired code rearrangement.

Reply:
Agree with lyakh that if possible this should be done in a way that makes it more reviewable. Can't the commit be split into smaller pieces that would show differences better ?

Reply:
I made additional preparation commit.